adafruit_httpserver/response.py

```diff
@@ -14,184 +14,206 @@
+    def send(
+        self,
+        body: str = "",
+        content_type: str = None,
+    ) -> None:
         """
-        Send the constructed response over the given socket.
+        Sends response with content built from ``body``.
+        Implicitly calls ``_send_headers`` before sending the body.
+
+        Should be called **only once** per response.
         """
+        if self._response_already_sent:
+            raise RuntimeError("Response was already sent")
 
-        if self.filename is not None:
-            try:
-                file_length = os.stat(self.root_path + self.filename)[6]
-                self._send_file_response(
-                    conn,
-                    filename=self.filename,
-                    root_path=self.root_path,
-                    file_length=file_length,
-                    headers=self.headers,
-                )
-            except OSError:
-                self._send_response(
-                    conn,
-                    status=CommonHTTPStatus.NOT_FOUND_404,
-                    content_type=MIMEType.TYPE_TXT,
-                    body=f"{CommonHTTPStatus.NOT_FOUND_404} {self.filename}",
-                )
-        else:
-            self._send_response(
-                conn,
-                status=self.status,
-                content_type=self.content_type,
-                headers=self.headers,
-                body=self.body,
-            )
-
-    def send_chunk_headers(
-        self, conn: Union["SocketPool.Socket", "socket.socket"]
-    ) -> None:
-        """Send Headers for a chunked response over the given socket."""
-        self._send_bytes(
-            conn,
-            self._construct_response_bytes(
-                status=self.status,
-                content_type=self.content_type,
-                chunked=True,
-                cache=self.cache,
-                body="",
-            ),
+        encoded_response_message_body = body.encode("utf-8")
+
+        self._send_headers(
+            content_type=content_type or self.content_type,
+            content_length=len(encoded_response_message_body),
         )
+        self._send_bytes(self.request.connection, encoded_response_message_body)
+        self._response_already_sent = True
 
-    def send_body_chunk(
-        self, conn: Union["SocketPool.Socket", "socket.socket"], chunk: str
+    def send_file(
+        self,
+        filename: str = "index.html",
+        root_path: str = "./",
     ) -> None:
-        """Send chunk of data to the given socket.  Send an empty("") chunk to finish the session.
+        """
+        Send response with content of ``filename`` located in ``root_path``.
+        Implicitly calls ``_send_headers`` before sending the file content.
+
+        Should be called **only once** per response.
+        """
+        if self._response_already_sent:
+            raise RuntimeError("Response was already sent")
+
+        if not root_path.endswith("/"):
+            root_path += "/"
```

Comment:
This maybe just me, but this creates different usages for the "/" between the route and the send_file command,
```
@server.route("/banner.png")
def base(request):
    hdr = HTTPHeaders({"Cache-Control": "max-age=31536000"})
    with HTTPResponse(request, headers=hdr) as res:
        res.send_file(root_path='/wwwroot/', filename='banner.png')
```
Note: the route needs the leading slash on the route file name to match, but when sending the file, the slash is at the end of the root_path, and not in the filename.

Reply:
Could you elaborate on that?
The `route` leading "/" is similar to Flask, that is it is a pathname similar to JS's `location.pathname`.
![image](https://user-images.githubusercontent.com/72110769/209758562-b86462c7-db8b-4a87-bc61-9a6f467daa8e.png)

The `send_file` leading and trailing "/" is optional, in your example all `wwwroot`, `/wwwroot` and `/wwwroot/` would result in the same, that is serving the file e.g. `H:\wwwroot\banner.png`. The only way I see it could be changed is by using `os.chdir()` somewhere in the code, that could mess up the relative paths.

Another thing is that in your example it might be better to use `root_path` in `HTTPServer.start()` for serving static files, thus eliminating the need for a `route` and handler.

Overall the `HTTPServer.start(root_path=...)` determines the directory where the server will look for files if **no handler is found**, so if a url is `http://192.168.0.2/wwwroot/banner.png` and there is not handler it wil look for `/<server_root_path>/wwwroot/banner.png` file in file system.

The `root_path` in `send_file` has the same purpose, but is not connected to server's `root_path`, that enables returning only selected files from non-shared directory, as in your example above e.g. '/wwwroot/style.css' is not accessible without a handler.

And finally the `filename` is simply the file's name, irrelevant to it's location in file system.

Reply:
Yes, I have the default root_path set in the server start line, and that does work.   However, I have to define a route handler for each of my files, since I want to add custom headers to the files.
```HTTPHeaders({"Cache-Control": "max-age=604800"})```

Anyway, I must have just changed too many things at once.

...as this works
```
@server.route("/css/wire.css")
def base(request):
    hdr = HTTPHeaders({"Cache-Control": "max-age=604800"})
    with HTTPResponse(request, headers=hdr) as res:
        res.send_file(root_path='/wwwroot/', filename='css/wire.css')
```
and so does this.
```
@server.route("/css/wire.css")
def base(request):
    hdr = HTTPHeaders({"Cache-Control": "max-age=604800"})
    with HTTPResponse(request, headers=hdr) as res:
        res.send_file(root_path='/wwwroot', filename='/css/wire.css')
```
the file operations do not care if there is a '//' in the path.

Reply:
Thanks for elaborating. Does this mean that it is working as expected fro you?

Also I belive it should be possible to reuse the same handler with many `route` decorators and then set `filename=request.path`, eliminating the need for multiple handlers and hard coding returned file name.

Reply:
Yes, everything is working as expected.

Thanks for the code savings tip.  That works fine too.